### PR TITLE
Explosion and Initialization Performance

### PR DESCRIPTION
**Pull request by adacovsk**
[performance]

Ever since I removed the sleeps and the splash screen once the nuke explodes, you observe and inward explosion in slow motion... It's very ugly and bad, so I decided to dive through some codebases to see what I could find. I'm not sure how to source or if there's licensing or anything regarding it.

I found this:
https://github.com/BeeStation/BeeStation-Hornet/pull/2874

## What this does
- Explosions go outward instead of inward
- Caches some data before the explosions, which are intensive
- removes some old admin code for modifying explosions to not use `explosion_block`? I don't think it's useful or very good
- uses different `spiral_block` code, I'm not sure if it's better, but it looks nicer, codewise.
- removes old unused `explosion_recursive` proc and `explosion_resistance`

## Why it's good
- faster, more beautiful explosions that explode outward
- faster initialization? maybe

## Changelog

:cl:
 * tweak: explosions look better, are faster, and explode outward (especially for nukes)

### CODE REVIEW

**Comment by djkramer123** [comment]
>Explosions go outward instead of inward

soulless

**Comment by Exxion** [comment]
> removes some old admin code for modifying explosions to not use explosion_block? I don't think it's useful or very good

IIRC this option predated the current explosion system and originally switched between circular explosions and *recursive* explosions, which were much more costly than the newer blocking method. Not sure why it was kept around

> Explosions go outward instead of inward

You are going to get lots of shit for this even though it's a good change and the actual creator of this explosion method has said that they only went inwards because he didn't know how to make them go outwards
It should probably be in the changelog, though

File: code/game/objects/explosion.dm [hunk 1]
```diff
@@ -215,3 +199,73 @@ var/explosion_shake_message_cooldown = 0
 
 	explosion_destroy_multi_z(epicenter, offcenter, devastation_range / 2, heavy_impact_range / 2, light_impact_range / 2, flash_range / 2, explosion_time)
 	explosion_destroy_multi_z(epicenter, offcenter, devastation_range / 2, heavy_impact_range / 2, light_impact_range / 2, flash_range / 2, explosion_time, whodunnit)
+
+/proc/CalculateExplosionBlock(list/affected_turfs)
+	. = list()
+	// we cache the explosion block rating of every turf in the explosion area
+	//explosion block reduces explosion distance based on path from epicentre
+	for(var/affected_turf in 1 to length(affected_turfs))
```
Comment (ghost) [inline code/game/objects/explosion.dm]:
```suggestion
	for(var/turf/affected_turf in affected_turfs)
```
Why not do this? Is it slower?

Comment (Exxion) [reply code/game/objects/explosion.dm]: That is slower, but `for(var/turf/affected_turf as anything in affected_turfs)` (or `var/turf/T` if you want) would be even faster than what's used here

EDIT: Well, I think, anyway. I'm actually not sure. Let me check

EDIT: Yes, the `as anything` version is the fastest, and by a larger margin than I expected

**Comment by jwhitak** [comment]
Explosions happening inward was a funny meme but ultimately I want performance in my funny space mans server.

File: code/game/objects/explosion.dm [hunk 2]
```diff
@@ -215,3 +199,73 @@ var/explosion_shake_message_cooldown = 0
 
 	explosion_destroy_multi_z(epicenter, offcenter, devastation_range / 2, heavy_impact_range / 2, light_impact_range / 2, flash_range / 2, explosion_time)
 	explosion_destroy_multi_z(epicenter, offcenter, devastation_range / 2, heavy_impact_range / 2, light_impact_range / 2, flash_range / 2, explosion_time, whodunnit)
+
+/proc/CalculateExplosionBlock(list/affected_turfs)
+	. = list()
+	// we cache the explosion block rating of every turf in the explosion area
+	//explosion block reduces explosion distance based on path from epicentre
+	for(var/turf/affected_turf in affected_turfs)
+		var/turf/T = affected_turfs[affected_turf]
```
Comment (Exxion) [inline code/game/objects/explosion.dm]:
```suggestion
	for(var/turf/T as anything in affected_turfs)
```
What you just changed this to is incorrect, and also would perform worse than what you had before if it were correct. This suggestion is faster than either.

**Comment by Eneocho** [comment]
How am I supposed to bombdodge if the blast comes from the inside.
What if the real explosions were the friends we made along the way?

**Comment by Exxion** [comment]
> How am I supposed to bombdodge if the blast comes from the inside.

By moving toward the bomb instead of away from it. It's otherwise the same.
Having it move outward makes it harder to bombdodge if you're near the bomb, but easier if you're far away.

File: code/modules/multiz/_HELPERS.dm [hunk 3]
```diff
@@ -101,37 +101,16 @@ var/global/list/visible_spaces = list(/turf/simulated/open, /turf/simulated/floo
 	var/list/spiraled_turfs = list()
```
Comment (ghost) [inline code/modules/multiz/_HELPERS.dm]: Remove this line, do other 4 things here too

File: code/modules/multiz/_HELPERS.dm [hunk 4]
```diff
@@ -97,41 +97,20 @@ var/global/list/visible_spaces = list(/turf/simulated/open, /turf/simulated/floo
...
-		for(var/i = 1, i < max_range, i++)
-			if(HasAbove(upturf.z))
-				upturf = GetAbove(upturf)
-				log_debug("Spiralling block of size [cube ? max_range : i + (max_range - i)] in [upturf.loc.name] ([upturf.x],[upturf.y],[upturf.z])")
-				spiraled_turfs += spiral_block(upturf, cube ? max_range : max_range - i, inward, draw_red)
-			if(HasBelow(downturf.z))
-				downturf = GetBelow(downturf)
-				log_debug("Spiralling block of size [cube ? max_range : i + (max_range - i)] in [downturf.loc.name] ([downturf.x],[downturf.y],[downturf.z])")
-				spiraled_turfs += spiral_block(downturf, cube ? max_range : max_range - i, inward, draw_red)
+	spiraled_turfs += spiral_block(epicenter,max_range,draw_red)
```
Comment (ghost) [inline code/modules/multiz/_HELPERS.dm]:
```suggestion
	. = spiral_block(epicenter,max_range,draw_red)
```

File: code/modules/multiz/_HELPERS.dm [hunk 5]
```diff
@@ -97,41 +97,20 @@ var/global/list/visible_spaces = list(/turf/simulated/open, /turf/simulated/floo
...
+	for(var/i = 1, i < max_range, i++)
+		if(HasAbove(upturf.z))
+			upturf = GetAbove(upturf)
+			log_debug("Spiralling block of size [cube ? max_range : i + (max_range - i)] in [upturf.loc.name] ([upturf.x],[upturf.y],[upturf.z])")
+			spiraled_turfs += spiral_block(upturf, cube ? max_range : max_range - i, draw_red)
```
Comment (ghost) [inline code/modules/multiz/_HELPERS.dm]:
```suggestion
			. += spiral_block(upturf, cube ? max_range : max_range - i, draw_red)
```

File: code/modules/multiz/_HELPERS.dm [hunk 6]
```diff
@@ -97,41 +97,20 @@ var/global/list/visible_spaces = list(/turf/simulated/open, /turf/simulated/floo
...
+		if(HasBelow(downturf.z))
+			downturf = GetBelow(downturf)
+			log_debug("Spiralling block of size [cube ? max_range : i + (max_range - i)] in [downturf.loc.name] ([downturf.x],[downturf.y],[downturf.z])")
+			spiraled_turfs += spiral_block(downturf, cube ? max_range : max_range - i, draw_red)
```
Comment (ghost) [inline code/modules/multiz/_HELPERS.dm]:
```suggestion
			. += spiral_block(downturf, cube ? max_range : max_range - i, draw_red)
```

File: code/modules/multiz/_HELPERS.dm [hunk 7]
```diff
@@ -97,41 +97,20 @@ var/global/list/visible_spaces = list(/turf/simulated/open, /turf/simulated/floo
...
 
 	return spiraled_turfs
```
Comment (ghost) [inline code/modules/multiz/_HELPERS.dm]: Remove this too

**Comment by adacovsk** [comment]
> You are going to get lots of shit for this

Nothing new 🤡

**Comment by WindowsErrors** [comment]
> > You are going to get lots of shit for this
> 
> Nothing new 🤡

not gonna lie but years without social interaction have left all of us with a warped perception of how we view the communities around us and society as a whole

**Comment by Kammerjunk** [comment]
can't wait to see the numbers on this
say, by dropping a 30/30/30 in the middle of the station

**Comment by adacovsk** [comment]
> can't wait to see the numbers on this say, by dropping a 30/30/30 in the middle of the station

I would guess about 2-3x faster for the nuke, but no hard calcs

**Comment by Kammerjunk** [comment]
imagine that

**Comment by Eneocho** [comment]
> > How am I supposed to bombdodge if the blast comes from the inside.
> 
> By moving toward the bomb instead of away from it. It's otherwise the same. Having it move outward makes it harder to bombdodge if you're near the bomb, but easier if you're far away.

If the bombs explode from the centre you can't reliably bombdodge if you are the bomber/a nearby victim. I want my funny survival when the traitor shartbombs next to me and I survive.

**Comment by ExpulsionSnooper** [comment]
lowers the skill ceiling 0/10

**Comment by Eneocho** [comment]
Requesting for the outward explosions to be their own separate PR and to keep this one strictly code optimizations.

**Comment by Exxion** [comment]
> Requesting for the outward explosions to be their own separate PR and to keep this one strictly code optimizations.

You realize that the optimizations themselves also make bombdodging harder, right? Also, you realize that bombdodging isn't an intentional feature, right?

**Comment by Eneocho** [comment]
>You realize that the optimizations themselves also make bombdodging harder, right? Also, you realize that bombdodging isn't an intentional feature, right?

I do. However, you should know that long time "features(TM)" eventually become features by virtue of their utility/user expectations/fun, like rocketjumping in tf2, nuclear gandhi in civ or deathbombing in touhou.
Bombdodging is a true feature and tradition at this point, making it harder via optimizations is just a skill issue and can be worked around with better timing, but changing the expected mechanic itself is a huge change considering the relatively common antag shardbombing.

**Comment by west3436** [comment]
> > You realize that the optimizations themselves also make bombdodging harder, right? Also, you realize that bombdodging isn't an intentional feature, right?
> 
> I do. However, you should know that long time "features(TM)" eventually become features by virtue of their utility/user expectations/fun, like rocketjumping in tf2, nuclear gandhi in civ or deathbombing in touhou. Bombdodging is a true feature and tradition at this point, making it harder via optimizations is just a skill issue and can be worked around with better timing, but changing the expected mechanic itself is a huge change considering the relatively common antag shardbombing.

a bug should never be seen as a feature dumbass

**Comment by Eneocho** [comment]
>a bug should never be seen as a feature dumbass

As I said, please direct your eyes to the examples I mentioned. They went from bug to feature. Hell, I can find you a list of bugs turned features.

**Comment by Eneocho** [comment]
https://tvtropes.org/pmwiki/pmwiki.php/Main/AscendedGlitch
Here you go, a list of bugs turned feature.

**Comment by Exxion** [comment]
> a bug should never be seen as a feature dumbass

This is false. However, this particular bug should not be.

**Comment by tuypo1** [comment]
rocket jumping is cool but bunny hops and straff jumps are cancer, just like bomb dodging

**Comment by Eneocho** [comment]
> rocket jumping is cool but bunny hops and straff jumps are cancer, just like bomb dodging

t. can't bunny hop, strafe jump or bombdodge

**Comment by Eneocho** [comment]
[Just believe.webm](https://user-images.githubusercontent.com/67024428/178052869-22b43671-c338-4034-bd8d-98539224932b.webm)
This is what they're trying to take from you.

**Comment by ShiftyRail** [comment]
noooo, my bomb dodging :(

**Comment by DeityLink** [comment]
this rewritten spiral_block() foregoes any checks of Z Level borders.

Also why the sleep(5) after each draw_red? makes the explosion check debug tool useless as you must wait 30 seconds for it to finish drawing.

**Pull request merged**